src/Symfony/Component/JsonPath/JsonCrawler.php

```diff
@@ -38,19 +38,34 @@ final class JsonCrawler implements JsonCrawlerInterface
         'value' => true,
     ];
 
+    /**
+     * @var resource|string|null
+     */
+    private mixed $raw = null;
+
     /**
      * @param resource|string $raw
+     *
+     * @throws \TypeError When the input is not a string or a resource
      */
-    public function __construct(
-        private readonly mixed $raw,
-    ) {
+    public function withRawData(mixed $raw): self
```

Comment:
what about this?

```suggestion
    public function fromJsonString($raw): self
```

Reply:
I was worried it might be a little too close to the vocabulary of factories, but I'm fine with this name

Reply:
Actually, `$raw` can be either a string or a resource, so `fromJsonString` is not perfectly accurate. We could either find a better name, or having both `fromJsonString()` and `fromJsonResource()`

Reply:
It's still accurate to me: the input remain a string, whether it's presented as an actual string or a stream

Which BTW makes me wonder: can one use the JsonPath against an associative array?

Reply:
> Which BTW makes me wonder: can one use the JsonPath against an associative array?

Currently no, but the support could be fairly easy to bring: `JsonCrawler` only `json_decode()` the raw data then the path is applied on the decoded array. So yes, by providing an array and skipping the decoding step, the whole logic remains the same.

Reply:
wouldn't it make sense then so support `fromJson(string|array|resource)`?

Reply:
I don't know. I find it a bit strange to have `fromJson()` which can directly take an array. But at the same time, it's technically so “quick” to add that it makes me doubt. 🙂 

I'm not sure I have a case in mind where someone would give a directly decoded array rather than the JSON string? Especially since, thanks to JsonStreamer, we're able to optimize the part of the string/resource to be decoded. Do you maybe have a use case in mind or is more out of curiosity?
Maybe a second implementation of CrawlerInterface could be added to support this different data structure actually (like Jérôme mentioned with BSON, etc.).

Reply:
Parsing a response from an API request done with HttpClient for example.

Reply:
Indeed, in this case I thought the person would rather pass on the raw answer, but I imagine it's a nice bonus for DX anyway without any maintenance cost. I've added support for `array`. Thanks 👍